### PR TITLE
Make connect() and connect_with() take a FnOnce for the codec

### PR DESCRIPTION
**Pull request by abreis**
This PR changes the signature of `serde_transport::tcp::connect()` to take a `FnOnce() -> Codec` for its codec parameter, which makes it symmetric to its `listen()`  counterpart:

```rust
let server_transport = tarpc::serde_transport::tcp::listen(addr, Json::default).await?;

// Can now be called with `Json::default` instead of `Json::default()`:
let client_transport = tarpc::serde_transport::tcp::connect(addr, Json::default).await?;
```

This prevents a minor footgun in which a user calling `connect()` with e.g. `Json::default` instead of `Json::default()` would get an inscrutable error from rustc.

### CODE REVIEW

File: tarpc/src/serde_transport.rs [hunk 1]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
     }
 
     /// Connects to `addr`, wrapping the connection in a TCP transport.
-    pub async fn connect_with<A, Item, SinkItem, Codec>(
+    pub async fn connect_with<A, Item, SinkItem, Codec, CodecFn>(
         addr: A,
-        codec: Codec,
+        codec: CodecFn,
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:157]:
```suggestion
        codec: impl FnOnce() -> Codec,
```

**Comment by abreis** [comment]
@googlebot I signed it!

File: tarpc/src/serde_transport.rs [hunk 2]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
...
         config: LengthDelimitedCodec,
     ) -> io::Result<Transport<TcpStream, Item, SinkItem, Codec>>
     where
         A: ToSocketAddrs,
         Item: for<'de> Deserialize<'de>,
         SinkItem: Serialize,
         Codec: Serializer<SinkItem> + Deserializer<Item>,
+        CodecFn: FnOnce() -> Codec,
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:165]:
```suggestion
```

File: tarpc/src/serde_transport.rs [hunk 3]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
     }
 
     /// Connects to `addr`, wrapping the connection in a TCP transport.
-    pub async fn connect_with<A, Item, SinkItem, Codec>(
+    pub async fn connect_with<A, Item, SinkItem, Codec, CodecFn>(
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:155]:
```suggestion
    pub async fn connect_with<A, Item, SinkItem, Codec>(
```

File: tarpc/src/serde_transport.rs [hunk 4]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
...
         Ok(new(
             Framed::new(TcpStream::connect(addr).await?, config),
-            codec,
+            codec(),
         ))
     }
 
     /// Connects to `addr`, wrapping the connection in a TCP transport.
-    pub async fn connect<A, Item, SinkItem, Codec>(
+    pub async fn connect<A, Item, SinkItem, Codec, CodecFn>(
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:174]:
```suggestion
    pub async fn connect<A, Item, SinkItem, Codec>(
```

File: tarpc/src/serde_transport.rs [hunk 5]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
...
         addr: A,
-        codec: Codec,
+        codec: CodecFn,
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:176]:
```suggestion
        codec: impl FnOnce() -> Codec,
```

File: tarpc/src/serde_transport.rs [hunk 6]
```diff
@@ -152,33 +152,35 @@ pub mod tcp {
...
     ) -> io::Result<Transport<TcpStream, Item, SinkItem, Codec>>
     where
         A: ToSocketAddrs,
         Item: for<'de> Deserialize<'de>,
         SinkItem: Serialize,
         Codec: Serializer<SinkItem> + Deserializer<Item>,
+        CodecFn: FnOnce() -> Codec,
```
Comment (tikue) [inline tarpc/src/serde_transport.rs:183]:
```suggestion
```

**Pull request merged**